### PR TITLE
Softmax vectorize

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 1]
```diff
@@ -12,13 +12,17 @@
 #include <ATen/cuda/NumericLimits.cuh>
 #include <type_traits>
 
+#include <ATen/native/cuda/Loops.cuh>
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/SoftMax.cu:15]: How is this file used?

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 2]
```diff
@@ -371,29 +375,40 @@ blockReduce(AccumT* smem, AccumT val,
...
 template <template<typename, typename> class Reduction, int ILP, typename T, typename AccumT>
 __device__ __forceinline__ AccumT
-ilpReduce(T* data,
+ilpReduce(int shift,
+          T* data,
           int size,
           const Reduction<T, AccumT>& r,
           AccumT defaultVal)
 {
+  typedef typename std::aligned_storage<ILP*sizeof(T), ILP*alignof(T)>::type LoadT;
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/SoftMax.cu]: Why are you using `aligned_storage` at some places while using `aligned_vector` else? IIRC, MSVC has some issue with `aligned_storage` that failed your vectorized dropout PR?

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 3]
```diff
@@ -371,29 +375,40 @@ blockReduce(AccumT* smem, AccumT val,
...
   AccumT threadVal = defaultVal;
   int offset = threadIdx.x;
 
+  // shift and do 1
+  if(shift > 0){
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/SoftMax.cu:399]: This is a smart idea of handling unaligned address!

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 4]
```diff
@@ -371,29 +375,40 @@ blockReduce(AccumT* smem, AccumT val,
...
-#pragma unroll
-    for (int j = 0; j < ILP; ++j)
-      tmp[j] = data[offset + j * blockDim.x];
+  for (; offset * ILP < (size - last); offset += blockDim.x) {
+    *value = reinterpret_cast<LoadT*>(data)[offset];
 
-#pragma unroll
-    for (int j = 0; j < ILP; ++j)
-      threadVal = r(threadVal, tmp[j]);
+    for (int j = 0; j < ILP; ++j) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/SoftMax.cu:417]: `#pragma unroll`?

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 5]
```diff
@@ -447,44 +474,50 @@ template <int ILP, typename scalar_t, typename accscalar_t, typename outscalar_t
...
+  using LoadT = at::native::memory::aligned_vector<scalar_t, ILP>;
+  using StoreT = at::native::memory::aligned_vector<outscalar_t, ILP>;
+
   extern __shared__ unsigned char smem[];
   auto sdata = reinterpret_cast<accscalar_t*>(smem);
   gradInput += blockIdx.x * classes;
   output += blockIdx.x * classes;
   gradOutput += blockIdx.x * classes;
 
+	const int shift = ((uint64_t)gradInput) % ALIGN_BYTES / sizeof(scalar_t);
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/SoftMax.cu]: nit: indent

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 6]
```diff
@@ -139,7 +143,7 @@ void SpatialSoftMax_getLaunchSizes(
 inline dim3 SoftMax_getBlockSize(int ILP, uint64_t dim_size) {
   uint64_t block_size = 1;
   uint64_t max_block_size = std::min(dim_size / ILP, static_cast<uint64_t>(max_threads));
-  while (block_size < max_block_size) block_size *= 2;
+  while (block_size < (max_block_size/2)) block_size *= 2;
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/SoftMax.cu]: Why there is a `/2` here?

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 1 above)
Comment (slayton58) [reply aten/src/ATen/native/cuda/SoftMax.cu:15]: This is needed for the following line `#include <ATen/native/cuda/MemoryAccess.cuh>` to compile successfully. `MemoryAccess.cuh` needs some values from `Loops.cuh`

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 2 above)
Comment (slayton58) [reply aten/src/ATen/native/cuda/SoftMax.cu]: Good point - this is code ported from an apex routine by Deyu - I missed that one

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 4 above)
Comment (slayton58) [reply aten/src/ATen/native/cuda/SoftMax.cu:417]: Good spot!

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 5 above)
Comment (slayton58) [reply aten/src/ATen/native/cuda/SoftMax.cu]: Sigh. tab vs. space

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 6 above)
Comment (slayton58) [reply aten/src/ATen/native/cuda/SoftMax.cu]: Checking on this..

**Pull request closed without merging**